exp/internal/controllers/machinepool_controller_noderef_test.go

```diff
@@ -199,3 +202,152 @@ func TestMachinePoolGetNodeReference(t *testing.T) {
 		})
 	}
 }
+
+func TestMachinePoolPatchNodes(t *testing.T) {
+	r := &MachinePoolReconciler{
+		Client:   fake.NewClientBuilder().Build(),
+		recorder: record.NewFakeRecorder(32),
+	}
+
+	nodeList := []client.Object{
+		&corev1.Node{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "node-1",
+			},
+			Spec: corev1.NodeSpec{
+				ProviderID: "aws://us-east-1/id-node-1",
+				Taints: []corev1.Taint{
+					clusterv1.NodeUninitializedTaint,
+				},
+			},
+		},
+		&corev1.Node{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "node-2",
+				Annotations: map[string]string{
+					"foo": "bar",
+				},
+			},
+			Spec: corev1.NodeSpec{
+				ProviderID: "aws://us-west-2/id-node-2",
+				Taints: []corev1.Taint{
+					{
+						Key:   "some-other-taint",
+						Value: "SomeEffect",
+					},
+					clusterv1.NodeUninitializedTaint,
+				},
+			},
+		},
+	}
+
+	testCases := []struct {
```

Comment:
Should we add a test case to test that an error is returned if either `AddAnotations` or `RemoveNodeTaint` returns false?

Reply:
yes! done